File: sample_project/typescript/astar.ts
import { Node } from './node';
import { Grid, Point } from './types';

// Interface definition
interface IAStarPathfinder {
    findPathTo(xend: number, yend: number): Node[] | null;
}

export class AStar implements IAStarPathfinder {
    private open: Node[] = [];
    private closed: Node[] = [];
    private path: Node[] = [];
    private now: Node;

    constructor(
        private maze: Grid,
        private xstart: number,
        private ystart: number,
        private diag: boolean
    ) {
        this.now = new Node(null, xstart, ystart, 0, 0);
    }

    // Arrow function method implementing interface method
    findPathTo = (xend: number, yend: number): Node[] | null => {
        this.closed.push(this.now);
        this.addNeighborsToOpenList(xend, yend);
        while (this.now.x !== xend || this.now.y !== yend) {
            if (this.open.length === 0) {
                return null;
            }
            this.now = this.open[0];
            this.open.splice(0, 1);
            this.closed.push(this.now);
            this.addNeighborsToOpenList(xend, yend);
        }
        this.path = [this.now];
        while (this.now.x !== this.xstart || this.now.y !== this.ystart) {
            this.now = this.now.parent!;
            this.path.unshift(this.now);
        }
        return this.path;
    };

    // Traditional method with function keyword
    private isInBounds(point: Point): boolean {
        return (
            point.x >= 0 && 
            point.x < this.maze[0].length &&
            point.y >= 0 && 
            point.y < this.maze.length
        );
    }

    // Arrow function as property
    private isWalkable = (point: Point): boolean => {
        return this.maze[point.y][point.x] !== -1;
    };

    // Method shorthand notation
    private addNeighborsToOpenList(xend: number, yend: number): void {
        for (let x = -1; x <= 1; x++) {
            for (let y = -1; y <= 1; y++) {
                if (!this.diag && x !== 0 && y !== 0) {
                    continue;
                }
                const newPoint = {
                    x: this.now.x + x,
                    y: this.now.y + y
                };
                if (x === 0 && y === 0) continue;
                if (!this.isInBounds(newPoint)) continue;
                if (!this.isWalkable(newPoint)) continue;
                const node = new Node(
                    this.now,
                    newPoint.x,
                    newPoint.y,
                    this.now.g,
                    this.distance(newPoint.x, newPoint.y, xend, yend)
                );
                if (
                    this.findNeighborInList(this.open, node) ||
                    this.findNeighborInList(this.closed, node)
                ) continue;
                node.g = node.parent!.g + 1;
                node.g += this.maze[newPoint.y][newPoint.x];
                this.open.push(node);
            }
        }
        this.open.sort((a, b) => a.f() - b.f());
    }

    // Arrow function with explicit parameter types
    private distance = (x1: number, y1: number, x2: number, y2: number): number => {
        return Math.sqrt(Math.pow(x2 - x1, 2) + Math.pow(y2 - y1, 2));
    };

    // Traditional function expression assigned to property
    private findNeighborInList = function(list: Node[], node: Node): boolean {
        return list.some(n => n.x === node.x && n.y === node.y);
    };
}

File: sample_project/csharp/AStar.cs
namespace AStarPathfinding
{
    public class AStar
    {
        private readonly List<Node> _open = new();
        private readonly List<Node> _closed = new();
        private readonly List<Node> _path = new();
        private readonly int[][] _maze;
        private Node _current;
        private readonly int _xStart;
        private readonly int _yStart;
        private int _xEnd, _yEnd;
        private readonly bool _diag;

        public AStar(int[][] maze, int xStart, int yStart, bool diag)
        {
            _maze = maze;
            _current = new Node(null, xStart, yStart, 0, 0);
            _xStart = xStart;
            _yStart = yStart;
            _diag = diag;
        }

        public List<Node>? FindPathTo(int xEnd, int yEnd)
        {
            _xEnd = xEnd;
            _yEnd = yEnd;
            _closed.Add(_current);
            AddNeighborsToOpenList();

            while (_current.X != _xEnd || _current.Y != _yEnd)
            {
                if (!_open.Any())
                    return null;

                _current = _open[0];
                _open.RemoveAt(0);
                _closed.Add(_current);
                AddNeighborsToOpenList();
            }

            _path.Insert(0, _current);
            while (_current.X != _xStart || _current.Y != _yStart)
            {
                _current = _current.Parent!;
                _path.Insert(0, _current);
            }

            return _path;
        }

        private void AddNeighborsToOpenList()
        {
            for (int x = -1; x <= 1; x++)
            {
                for (int y = -1; y <= 1; y++)
                {
                    if (!_diag && x != 0 && y != 0)
                        continue;

                    var node = new Node(_current, _current.X + x, _current.Y + y, _current.G, Distance(x, y));

                    if ((x != 0 || y != 0) &&
                        _current.X + x >= 0 && _current.X + x < _maze[0].Length &&
                        _current.Y + y >= 0 && _current.Y + y < _maze.Length &&
                        _maze[_current.Y + y][_current.X + x] != -1 &&
                        !FindNeighborInList(_open, node) &&
                        !FindNeighborInList(_closed, node))
                    {
                        node.G = node.Parent!.G + 1.0;
                        node.G += _maze[_current.Y + y][_current.X + x];
                        _open.Add(node);
                    }
                }
            }
            _open.Sort();
        }

        private double Distance(int x, int y)
        {
            return Math.Sqrt(Math.Pow(_xEnd - (_current.X + x), 2) + Math.Pow(_yEnd - (_current.Y + y), 2));
        }

        private bool FindNeighborInList(List<Node> list, Node node)
        {
            return list.Any(n => n.X == node.X && n.Y == node.Y);
        }
    }
}
